_stdlib/_types.dm

```diff
@@ -0,0 +1,92 @@
+#define childrentypesof(x) (typesof(x) - x)
+// consider declaring the base type abstract instead and using concrete_typesof instead of childrentypesof
+
+#define ABSTRACT_TYPE(type) /datum/_is_abstract ## type
+#define IS_ABSTRACT(type) text2path("/datum/_is_abstract[type]")
+/*
+usage:
+
+ABSTRACT_TYPE(/datum/atest)
+/datum/test
+/datum/test/child
+
+then you can do:
+for(var/type in typesof(/datum/test))
+	if(IS_ABSTRACT(type))
+		continue
+	do_stuff_with_type(type)
+
+Useful for parent types you don't want to be instantiated in places that use istype like this
+Note that the ABSTRACT_TYPE annotation *WILL* mark all predecesors of the type abstract too!
+That is a feature, not a bug.
+
+Turn on ABSTRACT_VIOLATION_CRASH or ABSTRACT_VIOLATION_WARN to get the server to yell at you
+when you instantiate an abstract type. CRASH will runtime, WARN will log it. Don't turn it on
+*permanently* on the live server (though turning it on once in a while is probably fine to see
+if there are any violations).
+*/
+#ifdef ABSTRACT_VIOLATION_CRASH
+/datum/New()
+	..()
+	if(IS_ABSTRACT(src.type))
+		CRASH("Attempt to instantiate abstract type '[src.type]'.")
+#elif defined(ABSTRACT_VIOLATION_WARN)
+/datum/New()
+	..()
+	if(IS_ABSTRACT(src.type))
+		logTheThing("debug", src, null, "Attempt to instantiate abstract type '[src.type]'.")
+#endif
```

Comment:
Alternative method of doing this would be:
```
#ifdef ABSTRACT_VIOLATION_CRASH
#define ABSTRACT_TYPE(T) /datum/_is_abstract ## T ;\
	T/New() { ;\
		if(IS_ABSTRACT(src.type)) CRASH("Attempt to instantiate abstract type '[src.type]'.") ;\
		..();\
	}
#endif
```
(and similarly for WARN). It would be less taxing on the server but if you did `ABSTRACT_TYPE(/obj/item/clothing/head)` that also marks `/obj/item/clothing` (and other predecesors) as abstract. However, you'd only get the warning / runtime on `/obj/item/clothing/head`.

Moreover if you did `ABSTRACT_TYPE(/obj/item/clothing/head)` and `ABSTRACT_TYPE(/obj/item/clothing)` you'd get duplicate warnings for `clothing`. That's why this is defined on `/datum/New` instead of being more specific.